feat(awscdk-construct): documentation (#179)
eladb/projen: docs/awscdk-construct.md

# AWS CDK Construct Library

This projen type is used for building AWS CDK constructs using the jsii. The [jsii](https://github.com/aws/jsii) allows 
you to write code once in Typescript and it will generate the Python, .net, and Java equivalents. Use this if you'd 
like to build and distribute CDK constructs for others to use.

> 

Reviewer: Add TOC

# Getting Started

Start like all projen projects:

```sh
npx projen new awscdk-construct
```

Review the resulting .projenrc.js file and make changes as needed. The following are some specific areas
you may want to set explicitly.

> A `README.md` file is not created but encouraged. It is required for publishing your construct for Python (see [jsii publishing](#jsii-publishing)).

Reviewer: Should we bootstrap one?

Author: I added a Readme class.

> ### Standard Node Module Fields

Reviewer: Why H3 instead of H2?

Author: missed some cleanup... fixed now.

These fields are your basic Node module setup:

```typescript
  authorAddress: "benisrae@amazon.com",
  authorName: "Elad Ben-Israel",
  description: 'Watching your CDK apps since 2019',
  name: 'cdk-watchful',
  license: "MIT",
  repository: "https://github.com/eladb/cdk-watchful.git",
  keywords: ["cloudwatch", "monitoring"]
```

All are pretty standard setup and nothing CDK-specific at this point. The `keywords` automatically gets 'cdk' so you don't
need to specify it. 

> ### CDK Dependencies

Reviewer: ```suggestion
### Dependencies
```

Please also add a section about non-CDK dependencies. Specifically, the difference between jsii deps and non-jsii deps (which require bundling)

Next are getting CDK dependencies added:

```typescript

>   entrypoint: 'lib/our-s3-bucket.js',

Reviewer: Not sure what's up here? The practice that I'd recommend is leaving this as the default `lib/index.js` and adding this to `index.ts`:

```ts
export * from './our-s3-bucket';
```

Author: 👍

  cdkVersion: "1.67.0",
  cdkDependencies: ["@aws-cdk/aws-ec2"],
  cdkTestDependencies: ["@aws-cdk/assert"],
```

The entrypoint can be set to point at your existing file, if not at `lib/index.js`. However, it's recommended that you

> rename your file to `index.js` and leave this un-set.

Reviewer: Remove

Author: 👍

`cdkDependencies` will add both dependencies and peerDependencies to your package.json file with a caret semver 
requirement (e.g. `^1.67.0`). You can set cdkVersionPinning to true to use a fixed version, but this means that any 
consumer of your library will have to use this exact CDK version. Likewise, `cdkTestDependencies` will add dependencies 
to the `devDependencies`. 

> 

Reviewer: Explain why both deps and peer deps are required (this is actually resolved by npm@7)

Author: I assume changes are needed if projen is to support the new npm@7 algorithm and not add the cdk deps to `dependencies`?

Reviewer: Yes, but it will take a while for users to migrate and this is a construct _library_, so we can't rely on people using npm@7.

Also, in CDK2.0, you'll only need a peer dependency.

Author: k. I don't believe there is anything to add to this until we decide npm@7 is adopted enough.

Additionally, you can add CDK dependencies using the methods:

```typescript
project.addCdkDependencies("aws-cdk/aws-sqs", "aws-cdk/aws-sns");
project.addCdkTestDependencies("aws-cdk/something-else");
```

The `@aws-cdk/assert` library is already added to the cdkTestDependencies for you.

### jsii Publishing

> As this is a [jsii](https://github.com/aws/jsii) constructs, it will cross-compile to other languages.  You can set up 

Reviewer: ```suggestion
As this is a [jsii](https://github.com/aws/jsii) project, it will cross-compile to other languages.  You can set up 
```

Author: 👍

any number of jsii target languages. 

```typescript
  dotnet: {
    dotNetNamespace: 'Acme.HelloNamespace',
    packageId: 'Acme.HelloPackage'
  },
  java: {
    javaPackage: 'com.acme.hello',
    mavenArtifactId: 'hello-jsii',
    mavenGroupId: 'com.acme.hello'
  },
  python: {
    distName: 'acme.hello-jsii',
    module: 'acme.hello_jsii'
  },
```

Each target requires a specific [Github project secret](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets) 

> to be setup:

Reviewer: Add a reference to [jsii-release](https://github.com/aws/jsii-release) which has detailed instructions on how to obtain these secrets.

Author: 👍

* npm - NPM_TOKEN

> * .net - NUGET_API_KEY

Reviewer: ```suggestion
* .NET - NUGET_API_KEY
```

Author: 👍

* Java:
  * MAVEN_GPG_PRIVATE_KEY
  * MAVEN_GPG_PRIVATE_KEY_PASSPHRASE
  * MAVEN_PASSWORD
  * MAVEN_USERNAME
  * MAVEN_STAGING_PROFILE_ID
* Python
  * TWINE_USERNAME
  * TWINE_PASSWORD
  
If you don't want to publish a particular package, do not include the `dotnet`, `java`, or `python` field.

### Workflows

Two workflows will be created as Github Actions:

* The Build workflow - controlled by the `buildWorkflow` field. On a 'pull_request' or 'workflow_dispatch' the construct
will be built and checked for anti-tamper (ensure no manual changes to generated files).
* The Release workflow - controlled by the `releaseWorkflow` field. On a push to `master` (overridden at

>  `props.defaultReleaseBranch`) the construct is built, anti-tampered, version bumped with a commit, pushed back to git, 

Reviewer: Replace the term "construct" across the doc with "construct library" or just "library"

Author: I think I got the right ones. Some places still seemed 'construct' was best.

 and then published to the configured artifact repositories (e.g. npm, pypi)

> Additionally, a [dependabot](https://dependabot.com/) file will be added unless `dependabot` is set to 'false'.

Reviewer: This should be under "Dependencies", not "Workflows".

### Construct Catalog

Finally, a field for setting up publishing to the [construct catalog](https://awscdk.io):

```typescript
  catalog: {
    announce: true,
    twitter: '@yourhandle'
  }
```

These values are optional but allow the construct catalog's Twitter account to mention your handle on the tweet. Setting `announce`
to 'false' will stop all tweets about the construct. However, the construct will still be indexed.

> ### docgen

Reviewer: ```suggestion
### API Documentation
```

Author: 👍

> # AWS CDK - Construct library

Reviewer: ```suggestion
# AWS CDK Construct Library
```

> This projen type is used for building AWS CDK constructs using the JSII.

Reviewer: Explain what's the value of using jsii and include some references to jsii docs.

jsii is spelled all lowercase BTW

Use this if you'd like to build and distribute CDK constructs for others to use.

> # Getting Started

Reviewer: why not use `projen new`?

Author: Yes, I'll add steps to use projen new instead of expecting this more manual approach.

Create a new instance of the project and set the appropriate fields.

```typescript
const project = new AwsCdkConstructLibrary({

>   authorAddress: "",

Reviewer: Instead of needing to maintain this, I rather we implement #168 and instruct people to use `npx projen new`

Author: 👍

    mavenGroupId: ''
  },
  python: {
    distName: '',
    module: '',
    twineRegistryUrl: ''
  },
 
  catalog: {
    announce: true,
    twitter: '@mattbonig'
  }
});
```

### Standard Node Module Fields

These fields are your basic Node module setup:

```typescript
  authorAddress: "benisrae@amazon.com",
  authorName: "Elad Ben-Israel",
  description: 'Watching your CDK apps since 2019',
  name: 'cdk-watchful',
  license: "MIT",
  repository: "https://github.com/eladb/cdk-watchful.git",
  keywords: ["cloudwatch", "monitoring"]
```

And will render out a package.json that looks like:

```json
  "name": "cdk-watchful",
  "description": "Watching your CDK apps since 2019",
  "repository": {
    "type": "git",
    "url": "https://github.com/eladb/cdk-watchful.git"
  },
  "author": {
    "name": "Elad Ben-Israel",
    "email": "benisrae@amazon.com",
    "organization": false
  },
  "keywords": [
    "cloudwatch",
    "monitoring"
  ],
  "license": "MIT",
```

All are pretty standard setup and nothing CDK-specific at this point. The `keywords` automatically gets 'cdk' so you don't
need to specify it. 

### CDK Dependencies

Next are getting CDK dependencies added:

```typescript

>   entrypoint: 'lib/index.js',

Reviewer: How is `entrypoint` related to CDK dependencies? (also, I don't think it needs to be set up)

Author: Totally forgot mentioning this. Since most people probably didn't name their construct's file 'main.ts' I figured this would need to be overridden. I'll make that clearer.

  cdkVersion: "1.67.0",
  cdkDependencies: ["@aws-cdk/aws-ec2"],
  cdkTestDependencies: ["@aws-cdk/assert"],
```

> With the above example you'll get a package.json that looks like this:

Reviewer: I don't think it's very useful to show the synthesized output for everything. Just describe the behavior. In this case, I would write something like:

`cdkDependencies` will add both `dependencies` and `peerDependencies` to your `package.json` file with a **caret** semver requirement. You can set `cdkVersionPinning` to `true` to use a fixed version, but this means that any consumer of your library will have to use this exact CDK version.

Author: Will do. I was putting the output because that usually helps me understand how properties transfer but I can remove it.

```json
  "main": "lib/index.js",
  "dependencies": {
    "@aws-cdk/assert": "^1.67.0",
    "@aws-cdk/aws-ec2": "^1.67.0",
    ...
  },
```

Additionally, you can add CDK dependencies using the method:

```typescript
project.addCdkDependencies("aws-cdk/aws-sqs", "aws-cdk/aws-sns");
```

> ### JSII Packaging

Reviewer: This is not just packaging, it's also publishing.

Author: But only if the related workflows are setup, right?

Reviewer: No, workflows are automatically created as soon as you specify `python`, `java` etc. No additional setup is required (besides installing the secrets to your github repo).

You can setup any number of JSII target languages. 

```typescript
  dotnet: {
    dotNetNamespace: 'Acme.HelloNamespace',
    packageId: 'Acme.HelloPackage'
  },
  java: {
    javaPackage: 'com.acme.hello',
    mavenArtifactId: 'hello-jsii',
    mavenGroupId: 'com.acme.hello'
  },
  python: {
    distName: 'acme.hello-jsii',
    module: 'acme.hello_jsii'
  },
```

> Will render out the following to the package.json:

Reviewer: Again, redundant. It also modifies the GitHub workflows.

```json
  "jsii": {
    "outdir": "dist",
    "targets": {
      "java": {
        "package": "com.acme.hello",
        "maven": {
          "groupId": "com.acme.hello",
          "artifactId": "hello-jsii"
        }
      },
      "python": {
        "distName": "acme.hello-jsii",
        "module": "acme.hello_jsii"
      },
      "dotnet": {
        "namespace": "Acme.HelloNamespace",
        "packageId": "Acme.HelloPackage"
      }
    },
    "tsc": {
      "outDir": "lib",
      "rootDir": "src"
    }
  }
```

If you don't want to publish a particular package, do not include the `dotnet`, `java`, or `python` field.

> 

Reviewer: In order to the github workflows to work, users need to add github secrets. This should be documented.

# AWS CDK Construct Library

This projen type is used for building AWS CDK constructs using the jsii. The [jsii](https://github.com/aws/jsii) allows 
you to write code once in Typescript and it will generate the Python, .net, and Java equivalents. Use this if you'd 
like to build and distribute CDK constructs for others to use.

# Table of Contents

* [Project Heirarchy](#project-heirarchy)
* [Getting Started](#getting-started)
  * [Standard Node Module Fields](#standard-node-module-fields)
  * [Dependencies](#dependencies)
  * [jsii Publishing](#jsii-publishing)
* [Workflows](#workflows)
* [Scripts](#scripts)
* [Construct Catalog](#construct-catalog)
* [API Documentation](#api-documentation)
* [Project structure](#project-structure)
* [Migrating existing projects](#migrating-existing-projects)
* [Feedback](#feedback)

# Project Heirarchy

AwsCdkConstructLibrary -> ConstructLibrary -> JsiiProject -> TypeScriptProject -> NodeProject -> Project

# Getting Started

Start like all projen projects:

```sh
npx projen new awscdk-construct
```

Review the resulting .projenrc.js file and make changes as needed. The following are some specific areas
you may want to set explicitly.

> ## Standard Node Module Fields

Reviewer: ```suggestion
## Module Metadata
```

These fields are your basic Node module setup:

```typescript

>   authorAddress: "benisrae@amazon.com",

Reviewer: Please normalize all quotes to single quotes (I know that `projen new` generates with double...)

  authorName: "Elad Ben-Israel",
  description: 'Watching your CDK apps since 2019',
  name: 'cdk-watchful',
  license: "MIT",
  repository: "https://github.com/eladb/cdk-watchful.git",
  keywords: ["cloudwatch", "monitoring"]
```

All are pretty standard setup and nothing CDK-specific at this point. The `keywords` automatically gets 'cdk' so you don't
need to specify it. 

## Dependencies

Next are getting CDK dependencies added:

```typescript
  cdkVersion: "1.67.0",
  cdkDependencies: ["@aws-cdk/aws-ec2"],
  cdkTestDependencies: ["@aws-cdk/assert"],
```

`cdkDependencies` will add both dependencies and peerDependencies to your package.json file with a caret semver 
requirement (e.g. `^1.67.0`). CDK dependencies must be both direct and peer dependencies, 

> see [this issue](https://github.com/aws/aws-cdk/issues/5064). You can set cdkVersionPinning to true to use a fixed 

Reviewer: ```suggestion
see [this issue](https://github.com/aws/aws-cdk/issues/5064). You can set `cdkVersionPinning` to `true` to use a fixed 
```

version, but this means that any consumer of your library will have to use this exact CDK version. 
Likewise, `cdkTestDependencies` will add dependencies to the `devDependencies`. 

Additionally, you can add CDK dependencies using the methods:

```typescript
project.addCdkDependencies("aws-cdk/aws-sqs", "aws-cdk/aws-sns");
project.addCdkTestDependencies("aws-cdk/something-else");
```

The `@aws-cdk/assert` library is already added to the cdkTestDependencies for you.

Additional dependencies can be added through `deps`/`dependencies` for jsii-based libraries or 
`bundledDeps`/`bundledDependencies` for libraries that are not built with the jsii. You can read more 
[here](https://github.com/aws/jsii/blob/master/docs/configuration.md#dependency-considerations).

A [dependabot](https://dependabot.com/) file will be added unless `dependabot` is set to 'false'.

## jsii Publishing

> As this is a [jsii](https://github.com/aws/jsii) project, it will cross-compile to other languages.  You can set up 

Reviewer: ```suggestion
As this is a [jsii project](./jsii.md), it will cross-compile to other languages.  You can set up 
```
